### ISSUE TITLE
LSH Ensemble for high dimensional sets

### ISSUE THREAD

**Issue by janM37** [ekzhu/datasketch#104]
Hi,
I've been reading your paper (http://www.vldb.org/pvldb/vol9/p1185-zhu.pdf) recently which I found really interesting. I wanted to try out the LSH Ensemble to see if it could have an impact on my project.
Following this page (http://ekzhu.com/datasketch/lshensemble.html) I've made up a very simple example here, with 9 sets containing 15 elements each, a query set of 5 elements and a very low threshold.
First results seem correct. Here is the sample code : 
```

#!/usr/bin/python3.6
import scipy as sp
import datetime
import random
from datasketch import MinHashLSHEnsemble, MinHash

class MyChrono :
    def __init__(self):
        self.t1=dict()
    def start(self,text):
        self.t1[text]=datetime.datetime.now()
    def stop(self,text):
        self.t2=datetime.datetime.now()
        delta=self.t2-self.t1[text]
        print(text+' seconds:'+str(delta.seconds)+'  microseconds:'+str(delta.microseconds)+'')
        return int(delta.seconds)

def compute_containment(s1,s2):
    s1 = set(s1)
    s2 = set(s2)
    containment = round(float(len(s1.intersection(s2)))/float(len(s1)), 2)
    return containment

def do_main(data):
    data_ids = list(data)
    chrono.start('minhashCreate')
    nrPerm = 256
    listIndexes = []
    lshensemble = MinHashLSHEnsemble(threshold=0.1, num_perm = nrPerm, num_part=64)
    for i in range(len(data_ids)-1):
        print("Now generating MinHash " + str(i+1) + "/" + str(len(data_ids)-1) + " for id " + str(data_ids[i]))
        m = MinHash(nrPerm)
        print("Current set has len " + str(len(data[data_ids[i]])))
        print(data[data_ids[i]])
        for d in data[data_ids[i]]:
            m.update(d.encode('utf8'))
        listIndexes.append((data_ids[i], m, len(data[data_ids[i]])))
    lshensemble.index(listIndexes)
    chrono.stop('minhashCreate')

    new_candidate = {'542532', '542531', '547837', '571925', '542534'}
    print("I will query : ", new_candidate)
    mn = MinHash(nrPerm)
    for d in new_candidate:
        mn.update(d.encode('utf8'))
    chrono.start('estimation')
    out = lshensemble.query(mn, len(new_candidate))
    chrono.stop('estimation')
    print("Estimation results : ")
    for key in out:
        print(key)
    print('\nGround truth results : ')
    chrono.start('groundTruth')
    ground_truth = list()
    for i in range(len(data_ids)-1):
        containment = compute_containment(new_candidate, data[data_ids[i]])
        ground_truth.append(containment)
        print("Containment of " + str(data_ids[i]) + " : "+ str(containment))
        print("Intersection : ", new_candidate.intersection(data[data_ids[i]]),'\n')
    chrono.stop('groundTruth')

if __name__ == '__main__':
    chrono = MyChrono()
    data = {1529329153: {'541133', '547837', '868815', '549309', '541136', '571925', '562111', '549300', '777663', '541138', '542535', '575254', '541132', '547839', '549302'}, 
            1529329160: {'541133', '547837', '868815', '549309', '541136', '571925', '562111', '549300', '541138', '542535', '575254', '541132', '547839', '549302', '549308'}, 
            1529333850: {'541133', '541134', '542531', '547837', '549303', '541136', '571925', '542532', '549300', '541138', '547835', '575254', '547831', '547839', '741865'}, 
            1529333852: {'607040', '541133', '541134', '542531', '547837', '541136', '571925', '542532', '549300', '541138', '547835', '542535', '575254', '547831', '547839'}, 
            1529333855: {'607040', '541133', '541134', '542531', '547837', '541136', '571925', '542532', '549300', '541138', '547835', '542535', '575254', '547831', '547839'}, 
            1529333857: {'607040', '541133', '541134', '542531', '547837', '541136', '571925', '542532', '549300', '541138', '547835', '542535', '575254', '547831', '547839'}, 
            1529333865: {'778282', '778280', '778281', '778285', '581606', '778288', '776533', '778284', '776532', '778289', '776531', '778287', '581604', '778283', '741865'}, 
            1529333872: {'607040', '541133', '541134', '542531', '547837', '541136', '571925', '542532', '549300', '541138', '547835', '542535', '575254', '547831', '547839'}, 
            1529333873: {'607040', '541133', '541134', '542531', '541136', '571925', '542532', '549300', '541138', '547835', '542535', '575254', '547831', '547839', '741865'}, 
            1529335140: {'542534', '541133', '542531', '547837', '541135', '541136', '571925', '542532', '549300', '541138', '547835', '542535', '575254', '547831', '547839'}}

    do_main(data)
```

I get the (summarized) following results : 
```
I will query :  {'571925', '542534', '542532', '542531', '547837'}
estimation seconds:0  microseconds:4
Estimation results : 
1529333857
1529329153
1529329160
1529333872
1529333873
1529333850
1529333852
1529333855

Ground truth results : 
Containment of 1529329153 : 0.4
Intersection :  {'571925', '547837'} 
Containment of 1529329160 : 0.4
Intersection :  {'571925', '547837'} 
Containment of 1529333850 : 0.8
Intersection :  {'542532', '571925', '542531', '547837'} 
Containment of 1529333852 : 0.8
Intersection :  {'542532', '571925', '542531', '547837'} 
Containment of 1529333855 : 0.8
Intersection :  {'542532', '571925', '542531', '547837'} 
Containment of 1529333857 : 0.8
Intersection :  {'542532', '571925', '542531', '547837'} 
Containment of 1529333865 : 0.0
Intersection :  set() 
Containment of 1529333872 : 0.8
Intersection :  {'542532', '571925', '542531', '547837'} 
Containment of 1529333873 : 0.6
Intersection :  {'542532', '571925', '542531'} 

groundTruth seconds:0  microseconds:487
```

Now suppose that I add a lot of other elements to initial data sets (for example, 20000 random integers to every set) : 

```
    for k in data:
        highD = set([str(k) for k in random.sample(range(500000,600000), 20000)])
        data[k] |= highD
    do_main(data)
```
Running the script again, the query output is now empty : 

```
I will query :  {'547837', '542531', '571925', '542534', '542532'}
estimation seconds:0  microseconds:4
Estimation results : 

Ground truth results : 
Containment of 1529329153 : 0.4
Intersection :  {'547837', '571925'} 
Containment of 1529329160 : 0.4
Intersection :  {'547837', '571925'} 
Containment of 1529333850 : 0.8
Intersection :  {'547837', '542532', '542531', '571925'} 
Containment of 1529333852 : 0.8
Intersection :  {'547837', '542532', '542531', '571925'} 
Containment of 1529333855 : 0.8
Intersection :  {'547837', '542532', '542531', '571925'} 
Containment of 1529333857 : 0.8
Intersection :  {'547837', '542532', '542531', '571925'} 
Containment of 1529333865 : 0.0
Intersection :  set() 
Containment of 1529333872 : 0.8
Intersection :  {'547837', '542532', '542531', '571925'} 
Containment of 1529333873 : 0.6
Intersection :  {'571925', '542531', '542532'} 

groundTruth seconds:0  microseconds:7385
```
I've observed that the more dimensions I add, the less results come into the output. Probably because the threshold gets too high?
Which finally leads to my question : how robust is supposed to be the algorithm to really high dimensions? Is there some heuristic to tune parameters?

Or maybe I'm missing something here...

Many thanks !

**Comment by ekzhu** [comment]
When you have an extreme difference between the query set size (5 in your experiment) and the indexed sets (>20000 in your experiment), the recall will be quite miserable. This is because the MinHash of your indexed sets are "overwhelmed" by all the random data you put in. LSH Ensemble aims to alleviate this issue by tunning the index to optimize for different query sizes, but this may still fail to accommodate the extreme difference in your case.

Usually, to increase recall you can increase `num_perm` and decrease `threshold`. However I doubt this will work in your example.

As for high dimension, MinHash-based techniques works very well for high dimensional data -- given the difference in size isn't too large.

**Comment by janM37** [comment]
Thanks for your answer !
Meanwhile, I've tested it again on 700 real data sets having from 1000 up to 250000 elements each, and performed a one element-query. This element was present in 98/700 sets.
I increased `num_perm` a lot (`=40000`) and put a really low `threshold` (`=0.001`).
Turns out the index worked pretty well, since 90 sets containing the desired element were in the output. I have here 100% precision and 91% recall. Not bad at all.

The 8 sets that were not recognized were typically really high dimensional sets (around 230000 elements each). I guess I would need to increase `num_perm` even more.

Considering those results, I'll investigate a bit more on even larger data sets. I'll let you know how this goes.
